tests/ut_notificationmanager/ut_notificationmanager.cpp

```diff
@@ -915,11 +941,82 @@ void Ut_NotificationManager::testRemoveRequested()
     LipstickNotification *notification1 = manager->notification(id1);
     connect(this, SIGNAL(removeRequested()), notification1, SIGNAL(removeRequested()));
 
-    QSignalSpy spy(manager, SIGNAL(notificationRemoved(uint)));
+    QSignalSpy removedSpy(manager, SIGNAL(notificationRemoved(uint)));
+    QSignalSpy closedSpy(manager, SIGNAL(NotificationClosed(uint,uint)));
     emit removeRequested();
     QCoreApplication::processEvents();
-    QCOMPARE(spy.count(), 1);
-    QCOMPARE(spy.last().at(0).toUInt(), id1);
+    QCOMPARE(removedSpy.count(), 1);
+    QCOMPARE(removedSpy.last().at(0).toUInt(), id1);
+    QCOMPARE(closedSpy.count(), 1);
+    QCOMPARE(closedSpy.last().at(0).toUInt(), id1);
+    QCOMPARE(closedSpy.last().at(1).toUInt(), static_cast<uint>(NotificationManager::NotificationDismissedByUser));
+}
+
+void Ut_NotificationManager::testImmediateExpiration()
+{
+    NotificationManager *manager = NotificationManager::instance();
+    uint id1 = manager->Notify("app1", 0, QString(), QString(), QString(), QStringList(), QVariantHash(), 0);
+
+    QHash<int, QVariant> expirationValues;
+    expirationValues.insert(0, id1);
+    expirationValues.insert(1, QDateTime::currentDateTimeUtc().toMSecsSinceEpoch());
+
+    qSqlQueryValues.clear();
+    qSqlQueryValues["SELECT * FROM expiration"].append(expirationValues);
+
+    QSignalSpy removedSpy(manager, SIGNAL(notificationRemoved(uint)));
+    QSignalSpy closedSpy(manager, SIGNAL(NotificationClosed(uint,uint)));
+    manager->MarkNotificationDisplayed(id1);
+    QCoreApplication::processEvents();
+    QCOMPARE(removedSpy.count(), 1);
+    QCOMPARE(removedSpy.last().at(0).toUInt(), id1);
+    QCOMPARE(closedSpy.count(), 1);
+    QCOMPARE(closedSpy.last().at(0).toUInt(), id1);
+    QCOMPARE(closedSpy.last().at(1).toUInt(), static_cast<uint>(NotificationManager::NotificationExpired));
+}
+
+void Ut_NotificationManager::testDelayedExpiration()
+{
+    NotificationManager *manager = NotificationManager::instance();
+    uint id1 = manager->Notify("app1", 0, QString(), QString(), QString(), QStringList(), QVariantHash(), 200);
+    uint id2 = manager->Notify("app2", 0, QString(), QString(), QString(), QStringList(), QVariantHash(), 100);
+
+    QueryValues expirationValues1, expirationValues2;
+    expirationValues1.insert(0, id1);
+    expirationValues1.insert(1, QDateTime::currentDateTimeUtc().toMSecsSinceEpoch() + 200);
+    expirationValues2.insert(0, id2);
+    expirationValues2.insert(1, QDateTime::currentDateTimeUtc().toMSecsSinceEpoch() + 100);
```

Comment:
worth testing one where the expiration time is prior to the current date time, just for sanity checking?